pilot/pkg/dns/monitoring.go

```diff
@@ -0,0 +1,49 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dns
+
+import (
+	"istio.io/pkg/monitoring"
+)
+
+func init() {
+	monitoring.MustRegister(requests)
+	monitoring.MustRegister(upstreamRequests)
+	monitoring.MustRegister(failures)
+	monitoring.MustRegister(requestDuration)
+}
+
+var (
+	requests = monitoring.NewSum(
+		"dns_total_requests",
+		"Total number of DNS requests.",
+	)
+
+	upstreamRequests = monitoring.NewSum(
+		"dns_total_upstream_requests",
+		"Total number of DNS requests forwarded to upstream.",
+	)
+
+	failures = monitoring.NewSum(
+		"dns_total_upstream_failures",
+		"Total number of DNS requests forwarded to upstream.",
+	)
+
+	requestDuration = monitoring.NewDistribution(
+		"dns_upstream_request_time_ms",
```

Comment:
cardinality may be an issue here @nrjpoddar @douglas-reid any guidance?

Reply:
Are these metrics going to be enabled by default when DNS proxy will be enabled or is it going to be opt in?

Reply:
It is not opt-in currently. Enabled when dns proxy is on.

Reply:
So if a user has enabled DNS_PROXY, then on upgrades they will see an increase in their metric storage after this change. Should we guard it and keep default off?

Reply:
yes. But since DNS proxy is still experimental and the amount of metrics we are adding here are not too many - I am guessing it should be fine. However, I am not against adding a flag defaulted to false, if you think it is better

Reply:
I will let others weigh in too but an opt-in option might be better

Reply:
adding 8 bucket distribution is like adding another istio standard distribution metric sans other labels. We can probably reduce it to 6. So I don’t think cardinality is a big issue. Since dns itself is experimental we can leave this in and create an option based on feedback.

we many need to systematically add metrics opt in for everything for the internal metric generation, but does not have to be here on an adhoc basis.

Reply:
Thanks I have reduced it to 6 buckets. @howardjohn can you PTAL?

Reply:
SGTM.

Reply:
Should we use seconds instead of ms? See https://prometheus.io/docs/practices/naming/#base-units